### DISCUSSION TITLE
$_LIBDIRFLAGS is ignored with TEMPFILE

### DISCUSSION THREAD

**Discussion by mtribiere** [SCons/scons#4347] [Q&A]
I'm using the following setup for TEMPFILE :

```
env["LINKCOM"]  = "$LD $LINKFLAGS -o $TARGET ${TEMPFILE('$SOURCES')} $_LIBDIRFLAGS $_LIBFLAGS"

env["TEMPFILE"] = SCons.Platform.TempFileMunge
env["TEMPFILEPREFIX"] = '@'
env["TEMPFILEARGJOIN"] = '\r\n'
``` 

However the `$_LIBDIRFLAGS` and `$_LIBFLAGS` part is getting ignored and never make it to the actual command, causing the build to fail. Example: 
- It should be:
```
ldppc -someFlag -o prog @rspFile.lnk -LIB_A
```

- But it is:
```
ldppc -someFlag -o prog @rspFile.lnk
```

Maybe there is something I'm missing here?

**Comment by bdbaddog** [comment]
You're using it wrong.
Here's how it's configured for MSVC
`    env['CCCOM']      = '${TEMPFILE("$CC $_MSVC_OUTPUT_FLAG /c $CHANGED_SOURCES $CFLAGS $CCFLAGS $_CCCOMCOM","$CCCOMSTR")}'`

It expects to have the whole command line, with the first arg being the program run which can be passed for example `${TEMPFILE(...)}`

so in your case

`ldppc @the_temp_file`

**Reply by mtribiere** [reply]
Alright I get that, however I need to have **only** the `.o` files in the temp file and the rest of the arguments in the command. Is that possible?

**Reply by bdbaddog** [reply]
So having the rest of the args in the command file won't work with that compiler?
Or is this your preference?

**Reply by mtribiere** [reply]
It just won't work at all, `ldppc` is very picky will only allow object files in the response file. That's why it's nececessary for my use-case to have the rest of the arguments in the command line

**Reply by bdbaddog** [reply]
Gotcha.
Can you put the other flags before the `${TEMPFILE(...)}`?
The problem is TEMPFILE currently returns.

```        cmdlist = [cmd[0], prefix + native_tmp + '\n' + rm, native_tmp]
```
So it should still have the remainder of your line, but it'd be after the newline and rm of the temp file.

**Reply by mtribiere** [reply]
Sadly no, it's arguments like : `-Ldependencies/LIB_A/lib` and `-lLIB_A`. So they need to be at the end of the command to link properly

**Reply by mwichmann** [reply]
> It just won't work at all, `ldppc` is very picky will only allow object files in the response file. That's why it's nececessary for my use-case to have the rest of the arguments in the command line

The GCC implementation went in somewhere in the 4.x series, and clearly treats the contents as options - meaning anything that could go on the command line, not just object file names (it's just a pure substitution, it looks like).  Your implementation must be different from that, sigh. 

https://gcc.gnu.org/legacy-ml/gcc-patches/2005-08/msg01443.html

**Reply by mtribiere** [reply]
Yeah the GCC I'm working with is a little older than that and has some modifications. I know, very picky but I don't really have the choice here.

**Comment by mwichmann** [comment]
Seems like you may want to write your own tempfile-handler, that uses some heuristic for what to put in the @file and what to leave on the line as it construct it.  `TEMPFILE` is set up to refer to `SCons.Platform.TempFileMunge` by default, but since `TEMPFILE` is a construction variable, you can point it to any function you like (`TempFileMunge` is actually a callable class).  Often wondered why this extra level of indirection was needed, but perhaps this is the reason...

**Reply by bdbaddog** [reply]
seems like the only path at the present time.
I'd add an additional argument (at least) for arguments after the ones which should go in the tempfile.
Might be able to mock something up later today.

**Reply by mtribiere** [reply]
How would you go about doing my own tempfile-handler in my situation?

**Reply by bdbaddog** [reply]
Copy the code for the existing TEMPFILE, modify it to your own needs then set 
`env['TEMPFILE'] = my_tempfile_implementation

**Reply by mtribiere** [reply]
That's exactly what I have been trying to do, but I cannot get it to work properly... Could you give me any clue?

**Reply by bdbaddog** [reply]
Would help if you explained what problems you're having.

**Reply by bdbaddog** [reply]
you might come join our discord server. It's a little more interactive for getting help.
https://discord.com/invite/bXVpWAy

**Comment by mtribiere** [comment]
If that helps any anyone, here is an example of how this can be archived:
``` python
import SCons.Subst
import SCons.compat
import tempfile

import importlib
import os
import sys

## This class is a handler to create a TEMPFILE
class TempFileRSP:
    def __init__(self, cmd, cmdstr = None):
        self.cmd = cmd
        self.cmdstr = cmdstr
    
    def __call__(self, target, source, env, for_signature):
                
        if for_signature:
            # If we're being called for signature calculation, it's
            # because we're being called by the string expansion in
            # Subst.py, which has the logic to strip any $( $) that
            # may be in the command line we squirreled away.  So we
            # just return the raw command line and let the upper
            # string substitution layers do their thing.
            return self.cmd
        
               
        # Now we're actually being called because someone is actually
        # going to try to execute the command, so we have to do our
        # own expansion.
        cmd = env.subst_list(self.cmd, SCons.Subst.SUBST_CMD, target, source)[0]

        # Check if the MAXLINELENGTH is reached and if that's
        # not the case, we just return the file sources
        try:
            maxline = int(env.subst('$MAXLINELENGTH'))
        except ValueError:
            maxline = 2048

        length = 0
        for c in cmd:
            length += len(c)
        length += len(cmd) - 1
        if length <= maxline:
            return self.cmd
        
        # Default to the .rsp suffix. We will append an 
        # .rsp suffix if none is given.
        if 'TEMPFILESUFFIX' in env:
            suffix = env.subst('$TEMPFILESUFFIX')
        else:
            suffix = '.rsp'

        if 'TEMPFILEDIR' in env:
            tempfile_dir = env.subst('$TEMPFILEDIR')
            os.makedirs(tempfile_dir, exist_ok=True)
        else:
            tempfile_dir = None

        fd, tmp = tempfile.mkstemp(suffix, dir=tempfile_dir, text=True) # <-- tempfile is a module here
        native_tmp = SCons.Util.get_native_path(tmp)

        if env.get('SHELL', None) == 'sh':
            # The sh shell will try to escape the backslashes in the
            # path, so unescape them.
            native_tmp = native_tmp.replace('\\', r'\\\\')
            # In Cygwin, we want to use rm to delete the temporary
            # file, because del does not exist in the sh shell.
            rm = env.Detect('rm') or 'del'
        else:
            # Don't use 'rm' if the shell is not sh, because rm won't
            # work with the Windows shells (cmd.exe or command.com) or
            # Windows path names.
            rm = 'del'

        if 'TEMPFILEPREFIX' in env:
            prefix = env.subst('$TEMPFILEPREFIX')
        else:
            prefix = '@'

        tempfile_esc_func = env.get('TEMPFILEARGESCFUNC', SCons.Subst.quote_spaces)
        args = [
            tempfile_esc_func(arg)
            for arg in cmd
        ]
        
        join_char = env.get('TEMPFILEARGJOIN', ' ')
        os.write(fd, bytearray(join_char.join(args) + "\n", 'utf-8'))
        os.close(fd)

        # Create the string to insert in the command based on the created rsp file
        # and the prefix given 
        cmdlist = [prefix + native_tmp]

        return cmdlist
```
